### PR TITLE
[Autograd/Checkpoint] Checkpoint implementation without reentrant autograd

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69060
* __->__ #69027

Resubmission of https://github.com/pytorch/pytorch/pull/62964 with the
suggestions and tests discussed in
https://github.com/pytorch/pytorch/issues/65537.

Adds a `use_reentrant=False` flag to `checkpoint` function. When
`use_reentrant=True` is specified, a checkpointing implementation that uses
SavedVariableHooks instead of re-entrant autograd is used. This makes it more
composable with things such as `autograd.grad` as well as DDP (still need to
add thorough distributed testing).

As discussed in https://github.com/pytorch/pytorch/issues/65537, the tests that we need to add are:

- [x] Gradient hooks are called once
- [x] works when input does require grads but Tensor that require grads are captures (like first layer in a nn)
- [x] works for functions with arbitrary input/output objects
- [x] distributed tests (next PR)

Note that this is only for `torch.utils.checkpoint`, if this approach overall looks good, we will do something similar for `checkpoint_sequential`. 

Differential Revision: [D32704467](https://our.internmc.facebook.com/intern/diff/D32704467/)

### CODE REVIEW

File: torch/utils/checkpoint.py [hunk 1]
```diff
@@ -275,3 +287,85 @@ def forward(input):
         input = checkpoint(run_function(start, end, functions), input,
                            preserve_rng_state=preserve)
     return run_function(end + 1, len(functions) - 1, functions)(input)
+
+
+class Checkpoint(torch.nn.Module):
```
Comment (rohan-varma) [inline torch/utils/checkpoint.py:292]: Implementation is entirely from https://github.com/pytorch/pytorch/pull/62964

File: test/test_autograd.py [hunk 2]
```diff
@@ -26,7 +26,7 @@
 from torch.autograd.profiler import (profile, record_function, emit_nvtx)
 from torch.autograd.profiler_util import (_format_time, EventList, FunctionEvent, FunctionEventAvg)
 import torch.autograd.functional as autogradF
-from torch.utils.checkpoint import checkpoint
+from torch.utils.checkpoint import checkpoint, Checkpoint
```
Comment (zhaojuanmao) [inline test/test_autograd.py:29]: nit: lint error

**Review by zhaojuanmao — COMMENTED** [review]
nice!

File: torch/utils/checkpoint.py [hunk 3]
```diff
@@ -142,7 +142,7 @@ def backward(ctx, *args):
         return (None, None) + grads
 
 
-def checkpoint(function, *args, **kwargs):
+def checkpoint(function, *args, use_reentrant: bool = True, **kwargs):
```
Comment (albanD) [inline torch/utils/checkpoint.py:145]: Maybe make it kwarg only? Or the fact that it is placed after `*args` already makes it kwarg only?

File: torch/utils/checkpoint.py [hunk 4]
```diff
@@ -166,9 +166,11 @@ def checkpoint(function, *args, **kwargs):
     be considered as part of autograd.
 
     .. warning::
-        Checkpointing currently only supports :func:`torch.autograd.backward`
-        and only if its `inputs` argument is not passed. :func:`torch.autograd.grad`
-        is not supported.
+        If ``use_reentrant=True`` is specified, checkpointing currently only
```
Comment (albanD) [inline torch/utils/checkpoint.py:169]: I feel like we could re-order some of the warnings. Especially because most of them only apply to the use_reentrant=True case.

File: torch/utils/checkpoint.py (hunk 1 above)
Comment (albanD) [reply torch/utils/checkpoint.py:292]: Why is this an nn.Module? This should be just a function no?

File: test/test_autograd.py [hunk 5]
```diff
@@ -4308,6 +4308,35 @@ def test_checkpointing(self):
...
+            nn.Linear(nz_inp, nz_bottleneck),
+            nn.ReLU(),
+            nn.Linear(nz_bottleneck, nz_inp)
+        )
+
+        feat_combined = []
+        for r in range(num_inp):
+            data_r = torch.empty(1, nz_inp)
+            data_r.uniform_()
+            data_r.requires_grad = True
```
Comment (albanD) [inline test/test_autograd.py:4332]: You can remove this to test the case where no input requires grad btw.

File: test/test_autograd.py [hunk 6]
```diff
@@ -4318,6 +4347,116 @@ def test_checkpoint_valid_reset_on_error(self):
...
+
+        self.assertEqual(b_grad, c_grad)
+        self.assertEqual(b_grad, d_grad)
+
+    def test_checkpointing_without_reentrant_dataparallel(self):
+        """
+        Verifies gradient correctness when checkpoint without reentrant autograd
+        is used in conjunction with DataParallel.
+        """
+        class ExpModule(torch.nn.Module):
```
Comment (albanD) [inline test/test_autograd.py:4378]: Should we use a Linear here maybe? So that the Module has some state?

File: test/test_autograd.py [hunk 7]
```diff
@@ -4318,6 +4347,116 @@ def test_checkpoint_valid_reset_on_error(self):
...
+            def __init__(self):
+                super().__init__()
+
+            def forward(self, inp):
+                return torch.exp(inp)
+
+        a = torch.randn(20, 20, requires_grad=True)
```
Comment (albanD) [inline test/test_autograd.py:4385]: nit: could be 2,2

File: test/test_autograd.py [hunk 8]
```diff
@@ -4318,6 +4347,116 @@ def test_checkpoint_valid_reset_on_error(self):
...
+        w = torch.randn(10, 10, requires_grad=True)
+        count = 0
+
+        def hook(grad):
+            nonlocal count
+            count += 1
+
+        w.register_hook(hook)
+        x = torch.rand(10, 10, requires_grad=True)
+        h = w * x  # Using w outisde the checkpoint
```
Comment (albanD) [inline test/test_autograd.py:4413]: outside

**Review by albanD — COMMENTED** [review]
Sounds mostly good!
I think we can improve the doc a little to make sure people are more inclined to use this new version.

File: torch/utils/checkpoint.py (hunk 1 above)
Comment (rohan-varma) [reply torch/utils/checkpoint.py:292]: It is indeed an nn.Module but it is basically called as a function in L223. Do you mean this should be a torch.autograd.Function?

The implementation is directly from https://github.com/pytorch/pytorch/pull/62964/files, do we want to refactor this to be a function or stick with the original implementation?

File: torch/utils/checkpoint.py (hunk 3 above)
Comment (rohan-varma) [reply torch/utils/checkpoint.py:145]: Yep, it is kwarg only since it is after *args: https://www.vegardstikbakke.com/python-keyword-only/

File: torch/utils/checkpoint.py (hunk 4 above)
Comment (rohan-varma) [reply torch/utils/checkpoint.py:169]: Moved this warning to the bottom then, and still kept the wording as use_reentrant=True is the default for now. Let me know if they could be structured better.

**Comment by rohan-varma** [comment]
@albanD Updated per your suggestions, although not completely sure about whether `class Checkpoint` should be a function or nn.Module. Could you take another look?

File: torch/utils/checkpoint.py (hunk 1 above)
Comment (albanD) [reply torch/utils/checkpoint.py:293]: I meant it could be a plain old python function.
This is not going to be public API anyways.

File: torch/utils/checkpoint.py [hunk 9]
```diff
@@ -190,6 +186,13 @@ def checkpoint(function, *args, **kwargs):
         model won't have gradients. At least one of the outputs needs to have
         :code:`requires_grad=True` as well.
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/utils/checkpoint.py:188]: Sounds good, I'll update this and add the appropriate tests as well

**Review by albanD — APPROVED** [review]
LGTM
Thanks for sending the PR!

**Comment by suo** [comment]
This unfortunately broke master (looks like a related test). https://hud.pytorch.org/commit/pytorch/pytorch/e032dae32904c6e90353e90051785514d6c5a7d9. I am reverting

**Comment by rohan-varma** [comment]
@suo Thanks for the revert, the failure makes sense and unfortunately it didn't run on the PR because it was marked as slowTest .
